### ISSUE TITLE
Holes to small

### ISSUE THREAD

**Issue by tmorris9** [alexrj/Slic3r#1588]
All versions of Slic3r up to and including 1.0

All holes be they round, square or in between slice too small. It's been reported on RepRap forum many times by users.

Same on Mac and Windows versions. Non machine specific, non object specific. All holes, openings print smaller than designed yet outer walls are +/- .01mm accurate.

Example a test piece with 16mm round hole 36 sided prints with an opening of 13.15mm. A 10mm hole prints with 9.5mm opening.

Kisslicer slices within the same tolerance as Slic3r does for outside tolerances.

**Comment by alexrj** [comment]
As of today, Slic3r has an option for compensating external size. More info at: http://manual.slic3r.org/troubleshooting/dimension-errors

**Comment by nicksears** [comment]
Alexrj, this issue needs to stay open. I've read every forum out there and this is definitely not resolved. You've detailed many of the compounding issues in the slic3r documentation, but EVERYONE still has this issue. X-Y compensation is nice but it shrinks the outside as well so it fixes one problem to make another. With this and the gap fill error in 1.2.9, this may be a nightmare for new users. I've been looking at the the arc compensation and doing some volume calculations for swept sold of different radii to determine what the best solution might be. All the other posts for reference:

http://forums.reprap.org/read.php?263,273929
http://forums.reprap.org/read.php?263,273929,274376#msg-274376
http://forums.reprap.org/read.php?263,134103,137818#msg-137818
https://github.com/alexrj/Slic3r/issues/1613
https://github.com/alexrj/Slic3r/issues/491
https://github.com/alexrj/Slic3r/issues/426
https://github.com/alexrj/Slic3r/issues/418
https://github.com/alexrj/Slic3r/issues/56
And lastly http://hydraraptor.blogspot.com/2014/06/why-slicers-get-dimensions-wrong.html?m=1

We need to get this fixed. Can anyone say they get holes within 0.25mm without compensation of one sort or another? I have a well calibrated printer that gets within 0.03mm OD, but is almost always 0.4-0.5 under in ID.

**Comment by nophead** [comment]
See http://hydraraptor.blogspot.co.uk/2011/02/polyholes.html for why round
holes come out too small and how to get them correct. Also note that if you
using multiple outlines you need to do the inner one first for a hole and
the outer one first for a perimeter.

**Comment by nicksears** [comment]
Hi nophead, thanks for responding! I've read your info on polyholes and it's a neat solution to the problem. Shaping a polygon around a hole mitigates the constant "smushing" towards the center that is caused by curving the extrudate in an arc. A polygon still has some of these effects in the corners, essentially turning them into fillets, but they don't affect fit since they're outside of the cylindrical hole we're trying to create. It's elegant that it has the relationship between the diameter and number or sides for the polygon, but it doesn't FIX the issue. It's more elegant than designing in an extra 0.5mm to the diameter of a hole, but it can get complicated when trying to create a thread or something more complex.

After addressing (or coping) with all of the other mechanical, processing, and material issues (I feel they're well summed up [here](http://manual.slic3r.org/troubleshooting/dimension-errors) and other places) we are still left with the issue of the volume mismatch between the inner and outer half of the revolved "[stadium](https://www.google.com/search?q=stadium+shape&oq=stadium+shape&aqs=chrome..69i57j0l5.1504j0j7&sourceid=chrome&es_sm=93&ie=UTF-8)" shape that makes up our perimeters.

The first half of the issue is the compounding of the error resulting from multiple perimeters. This is mitigated reasonably well by placing outer perimeters first. Since the compounding error is now pushed inwards, I think this means that the perimeter overlap could be lower, but I have not proven this. But even a single line will "smush" towards the center (not just "dragging" or shrinking) because of the flow math. 

However, to fix the other half of the issue, Slic3r should compensate for the fact that more material is deposited on the inside of an arc, possibly by offsetting the centerline based on the radius and possibly other geometric factors. This would fix the issue with a single circular perimeter, but it would probably fix the compounding issue with multiple perimeters as well. The solution is to figure out the ratio of the volume of the inner half of the path as a function of the radius. I'm no math major but I'm working it out. Hopefully I'll be back with an answer soon, because this is driving me crazy.

**Comment by nophead** [comment]
Dr Adrian Bowyer's maths for compensation for less flow needed inside can be found in RepRap wiki by following the link in my article. However, I think the dominant shrinking is caused by the drag towards the center, which depends on nozzle size and plastic viscosity, so would need an empirical solution.

**Comment by arildj78** [comment]
Even though there have been several attempts to describe the math behind this problem, I don't think any of the solutions will give a magic solution to the problem. Circles represented by polygons are surely one factor, dragging the extrusion another. How much each factor contributes to the total error is bound to be design dependent and probably even specific to each individual printer.

I believe we can get a good approximation with a simple function and trial and error to set the constants. The best would of course be a semi automatic calibration, where measures of a testprint can be plugged into the software to figure out the constants.

I suggest examining f(x)=ax^2+bx+c where x is the design radii and f(x) is the corrected radii. a, b and c are constants to be figured out by trial and error.

The difficult part of this may be detecting a hole and modifying it with a new radius.

**Comment by bstott** [comment]
For hole diameter ---- use the ratio of the design diameter to printed diameter times the designed diameter for input to the printer as the new radius for the printed hole and subsequent holes. Yes?  This would be the fudge factor for the specific printer and present calibration. Yes?

Example:
1. Print test hole.
2. Measure hole printed diameter.
3. Adjust per: (Design_diameter / Printed_diameter) X Design_diameter OR Design_diameter squared divided by Printed_diameter for the input diameter to print on the specific printer. 
4. Adjust design for resulted hole size for stl and slicing - OR - ???

How to put into --- firmware or splicer?

**Comment by nicksears** [comment]
The thing is, if we get too empirical about this, the solution won't stick if any parameters change significantly. That said, I'd be fine with separate inner/outer compensation (at least that might work better than xy size compensation. I think the most meaningful thing we can do is try and fix what we know is mathematically incorrect, and we can tackle compensating for other known issues later. 

I'm blaming the flow math around curves because it's a definite cause of inaccurate hole sizes. This is true even with a single perimeter on a perfectly rigid machine with perfect filament. If we fix this issue, it may not fix the entire issue with hole sizes, but it should fix the issue with multiple perimeters (we should be able to do them in any order) and it will likely make many aspects of printing more accurate.

**Comment by arildj78** [comment]
(Causes are cut'n paste from my entry on this issue 13 Mar 2014)
POSSIBLE CAUSES
----------------------------------
1 - Shrinkage
2 - More material on inside when turning
3 - Segmented polylines
4 - Filament cutting corners
5 - Z Wobble
6 - Low quality filament

Solution
------------------------------------
I don't know how to write code to correct for these. I will however help with the math if needed. Error 4 is possibly difficult to predict and will probably be a fudge factor of some sort. Error 5 and 6 will also be a fudge factor, either separate or included in error 4.

Does anyone here contribute to the code and are able/willing to do something about this?

**Comment by nicksears** [comment]
I think #1 should be ignored for now because it's a smaller/less common issue. Also, we should be able to either scale or "shrink" in X-Y to fix this (one thing that should be an easy change would be to have mm OR % in the XY compensation)

#2 is the issue I'm wrestling with. I think we need something like arc compensation put back in. I'm going to go back to the old version of slic3r that still has it to see how it acts. Anyone know when it was removed? I think I read ~v0.8 or so. Also, I've started evaluating toolpaths made in commercial "catalystEX" software by stratasys to try and figure out what offset they're using. I'll have more on this soon once I extrapolate nozzle size (or at least extrusion width) and see how it changes with hole/edge size. 

I dont feel like #3, segmented polylines should have much of an impact since, as nophead's post points out, only 22 lines are needed to reduce shrinkage to 1%. I can like with a 2.97mm 3mm hole.

I think with #4, filament cutting corners, this described wrong in the slic3r manual. Either this means filament dragging towards the center, and in that case I'm having a hard time coming up with an actual mechanism of how that's happening. Or, this is actually cutting corners, which is the increased flow towards the inside that we get in issue #2.
 
#5 and 6 are mechanical issues and while it might be good to have the ability to compensate for z wobble and crappy filament, it should be separate. You can't complain about 0.5mm hole variance with a >5% variance in filament and/or z wobbling ~0.5mm.

**Comment by nicksears** [comment]
OK I already have some initial data. I've analyzed quite a few simple files created by CatalystEX. (I don't know if this is necessarily free software, but it's not exactly useful unless you have a stratasys printer...) I got it from some university webpage so it's out there and not hard to find. I made a range of holes (inner) and posts (outer) in square and circular form, and sliced them with the software at 0.1" (0.196mm) layers for a "dimension elite", and it exports a .cmb file. I then opened it with the cmbviewer (within the program files folder) and measured the diameter of the inner or outer square or circular toolpath (click and drag) to see if they offset the perimeter for holes vs posts vs straight edges and I can say they most certainly do NOT. While I did not know the nozzle diameter, I found the extrusion width to be around 0.5mm and the measurements were always the exact number +/- the nozzle diameter (depending if it was an inner or outer hole. This did decrease for lower layer heights to about 0.4 (I think they use a ~0.35mm nozzle).

**Comment by wentbackward** [comment]
Hi Nick,
I have a Stratasys uPrint SE and they temp runs close to 300'C when printing, the filament is clearly far more liquid and the ABS is of a special formulation geared to their temperature and to the extrusion process. ... I think this is fortunate as I would not use any knowledge about a commercial process in an open source product unless you specifically know the legalities. The Uprint SE is remarkably accurate as well as being of heavy industrial construction, surprisingly vibration still sneaks in, but is kept to a minimum. I don't feel a specific process for a specific machine with specific materials is going to be a fruitful exploration.

You'll kick yourself but from your previous msg 0.1" is 2.54mm (0.1 x 25.4) :).

Good luck, this is definitely a required feature! I can try to provide help where I can, but unfortunately not using any open source printers right now.
Paul

**Comment by arildj78** [comment]
Is this a problem with other open source slicers or toolchains? As far as I can see this issue was closed quite a while ago, and I'm considering abandoning slic3r in favor of a software that don't have this problem (if something like that exists).

If we have someone here with the skills to make changes to the source, then I'll be glad to contribute. Final goal must be to be able to print a post and a hole of equal diameter that fits together.

**Comment by xoan** [comment]
> Final goal must be to be able to print a post and a hole of equal
diameter that fits together.

It is theoretically impossible, I think.

A point in space doesn't take space, but two points in the space placed in
the same coordinates are, actually, the same point.

So the inner contour of the hole and the outer contour of the post can't
have the same dimension, because in that case, a point in the hole contour
have a coincidental point in the post contour (all of this without taking
account of material tolerances and all the things described in Slic3r
manual about Z errors like misaligned perimeters).

**Comment by wentbackward** [comment]
@xoan et al.

it is not appropriate to specify only the diameter of a circle with no mass. A pipe for example is generally described with an outer diameter (OD) and a wall thickness. A ball bearing can be described by it's inner diameter and outer diameter (along with other things). So you CAN build a model in 3D CAD representing such parts as pipes, bearings, solids with holes and you can precisely describe the dimensions such as an inner diameter.

What is more difficult to describe are the tolerances, but all this has been thought of for us too and supported within many CAD systems. For example circularity is a geometric characteristic that described how round something is and one of the ISO GT&D characteristics standardised for communicating compliance / tolerances to a standard.

Finally when melting plastic we need to carefully consider the coefficient of expansion of the material (people refer to as 'shrinkage'). Again all this has been thought of and carefully worked out by people far cleverer than I'll ever be. 

I mention all this here to give the correct terms for willing slic3r programmers without engineering knowledge to be able to google the right things to understand how the eventual calculations need to be made.
- Paul

**Comment by ioquatix** [comment]
I'd like to see progress on this issue. I just printed a hole with d=1.5 and it came out  <0.5mm. I feel like the model for how plastic is deposited might be the culprit as it's probably likely that the plastic is smushing out further than slic3r might anticipate. I tried the XY compensation but it looked like it generated slightly buggy output.

**Comment by lordofhyphens** [comment]
Something that massive feels more like overextrusion than anything else the
slicer might be doing. Does the problem follow different settings or change
of slicers (like cura or Craftware)?

**Comment by DrLex0** [comment]
This is definitely still an issue, but if the length of this thread wasn't enough of an indication: there is no obvious cause.

I printed an object that has some internal holes of 3 and 6mm diameter, as well as one hole half-open to the outside, with an 11.25 mm diameter. The internal holes ended up about 0.4mm too small in diameter. The outside hole was printed accurately to almost 0.01mm.

So, to test the hypothesis that outside perimeters are treated differently from inside, I created a cheese-like test object, with exactly the same situation: two internal holes and two holes open to the outside, 5 and 8 mm diameters. Hypothesis debunked: all holes ended up printed too small. The degree of mismatch only seems to depend on diameter. The 8mm hole is about 0.3mm too small, the 5mm one about 0.4mm. Maybe the equation used by Slic3r ignores a certain effect that only manifests itself at the smallest diameters.

If there is no fundamental solution, it would be nice to split up XY size compensation between outside and inside perimeters, so we can independently tweak those depending on the model.

**Comment by mrbi11** [comment]
I believe this may be a subtle effect of the slight miscalculation (one 
could argue miscalibration) of the amount of material
being laid down. Recall you are not printing, your are squirting a round 
tube of liquid down. It needs to be slightly larger than
the square it is supposed to fit in, so the walls of each liquid thread 
become squished roughly flat against each other.
As the printing occurs usually outside to inside, the accumulation is to 
be shifted inwards. It cant really be "fixed" by
reducing the amount of liquid being laid down, or the threads stop 
deforming to adhere to each other and it de-laminates.

So I dont have any uniform solution to offer, since I expect it varies 
from printer to printer.
As a practical matter, I would print the thing with the hole and either 
drill it to size when that is possible,
or do what I did on several pieces, measure the wrong sized hole, then 
reprint with a larger hole size calculated to correct the error.
Neither particularly ideal.

I would note that when printing very small holes, reversing the order of 
printing may make it worse.
Why? because you are dragging a liquid stream. If you sharply turn a 
corner, the liquid is viscose enough
to be dragged along to the side if one turns a sharp corner. I expect if 
you make any sharp turn the actual effect
is that the turn becomes a small radius curve. Unless the printer 
stopped, disengaged till it cooled, then reengaged
any very fast change of direction the fluid will tend to drag along. 
Basically for really small holes printing inside to outside
may just make it get pulled to the side instead of pushed to the side.

Good luck to all.

*peace,
bill kelley
512 266 1896 *

**Comment by mrvn** [comment]
@DrLex0 How did you design the holes? What is the actual radius of the polygon and how many facets does it have (for each hole)? It sounds like you aren't correcting the polygon to be an outcircle.

@mrbi11 That idea does not track. Holes are printed as perimeters and either from the inside out (meaning outside in for holes) or outside first depending on your settings. Default is inside out.

As for draging the filament that is indeed the unsolved problem. But the rule for holes, as implemented in the polyholes openscad library) is to use less faces the smaller the hole and to make the polygon stay outside the hole. The drag effect seems to usualy cancel out the polygon being to big and give near perfect holes. But that depends on the printer speed, filament and temperature too. It just happen to accidentally work out about right for most people.

**Comment by DrLex0** [comment]
I used 64 facets for the holes in this ‘cheese test,’ which means the effect of the hole being approximated by a polygon should be negligible considering the small radii. I guess I'll try polyholes next time.
It would be great if we could just design models with “a hole of this diameter here” and the slicer would render this in the best way for the specific printer, instead of having to design the model with limitations of the printer in mind…

**Comment by whosawhatsis** [comment]
@DrLex0 If you're not doing the 1/cos(180/n) math, where n is the number of segments in the "circle", then you are actually specifying a hole that is smaller than the circle you want to fit inside. That's a limitation of the design software, and polyholes is a workaround for it.

**Comment by DrLex0** [comment]
@whosawhatsis Yes, but for 64 segments, the error of not compensating amounts to 0.12%, not the +8% I measured in my test.

**Comment by mrvn** [comment]
@DrLex0 Slic3r can not render the hole since the stl file contains the hole already rendered. You have to do the hole right in your design software.

Another effect you have with 64 segments on a 3mm hole is that the segments get really short. So short that, at least when printing over serial, the time to transmit the gcode is larger than the time to move the head. That causes a stutter on the printer where the head moves, stops, moves, stops, moves, stop. Can't be good for print quality.

**Comment by DrLex0** [comment]
@mrvn My point is that if we need to design holes with limitations in mind of the particular printer they will be printed on, then there is not even a way to “do the hole right” if the goal is to publish a model that anyone can download and print on any printer, including ones where there is not even a piece of machinery that traces contours. We shouldn't be limiting models to jaggy holes just because a certain type of printer will only print those jaggy holes correctly.

Of course, there is indeed no fundamental solution for this as long as we keep on using representations that consist of pre-rendered surfaces (heck, we shouldn't even be representing surfaces, but volumes). At best, there can be workarounds, but those workarounds should not be performed during the modelling stage. A 3D model should be like a technical drawing, and a key requirement of a technical drawing is that it is independent of the particular way in which the part will be manufactured.

**Comment by nophead** [comment]
Polyholes work on any printer that gets straight lines in the right place.
They may look a bit faceted on a high resolution printer but they work fine
to hold a fastener or a shaft. In the former case they get covered up
anyway. Since I started using them I have no problem with hole sizes.

**Comment by mrbi11** [comment]
While your statement is true, I would tend to question whether any 
melted filament printers are capable of laying down
straight lines where the head passes.

The filament must be laid down slightly oversized, and is a fluid. This 
may be a tiny error, perhaps 1/100 of a trace width.
But to lay a solid wall, say the trace width is 0.5mm and one wishes to 
lay a solid wall only 5mm thick, thats 10 traces
and the cumulative push of fluid to one side is now 1/10 of a trace or 
.05 mm x 2 for a hole. I think in fact this is an extremely
optimistic with 1/100 and this would mean the round liquid filaments 
barely touch and this would de-laminate.
This is one of the reasons its basically not possible to print large 
solid pieces at all, one must have fill areas with
some space in them. (Its not the only reason, as the dimensions also 
change a lot during cooling) A more realistic
case is the filament is 1/10ths to 1/5th oversized to get "good" 
deformation into the rectangle filament one really wants.
Then the same example is .5 to 1mm x2 for a hole printed from outside in.

Printing from inside out gets a different problem I mentioned last time, 
the filament is molten still as the head changes
direction, so is pulled to the side as the heat changes direction. I 
always got more repeatable results from outside in.
Otherwise the tinier the hole the more likely it just disappears.

Either way, melted filament printers are not capable of precise 
placement of the filaments, but are always subject to drag
and displacement.

There are papers out there describing this, I can't take credit for 
being the first to consider this, rather I'm describing
(probably not well) issues that are known limitations with the methodology.

If you think not, try "printing" with a toothpaste tube. Do it on graph 
paper so you can measure the displacement.
Make sure the traces have nice flat walls, not barely touching. Try 
turning a tight corner without dragging it along.
I bet you always get slightly curved corners. Mark some straight traces 
side by side before you lay a toothpaste "wall" down.
If you lay it down trace after trace, you will see the traces creep to 
the side.

Best wishes.

*peace,
bill kelley
512 266 1896 *

**Comment by nophead** [comment]
I always print with a single outline to get accurate dimensions. If I want
it strong I use 95% infill. So there is no displacement, only drag around
corners. Polyholes avoid the drag problem by having as few corners as
possible so that they are far enough outside the circle to not impinge on
it. That way my STL files are printer independent.

To print perfecting round circles accurately you would need a different
file representation for circles and then filament type / nozzle size /
print speed / layer height / etc dependent slicing.

**Comment by nicksears** [comment]
Stratasys (even a cheap mojo) gets holes within 0.1mm). You need a 3mm hole? Design a 3mm hole. That's how it should be. My well calibrated reprap is typically 0.3-0.4. Poor z rods, filament, all the stuff mentioned on [http://manual.slic3r.org/troubleshooting/dimension-errors](url) are valid, but don't fully solve the problem.

Polyholes is a unique solution, but it's more an indication of the problem. It's a design constraint that you need to export an stl with quality equal to or beyond what you need. All this talk about having too many facets, or needing to design it with so many facets is not necessary. With polyholes, we're essentially designing a bigger hole by circumscribing it within a polygon. 

What we really need to do is modify flowrate in an arc depending on the radius. I've visually checked the gcode (line view) from stratasys machines and they have no compensation in terms of the actuall coordinates. For a .5mm nozzle and 3mm hole, the line has a diameter of .5mm (offset by half the diameter on both sides). There are other issues with the outer edge of the extruded line being undersized (in general, but also if we reduce flow) but i'm looking into how they handle this to get some ideas.

**Comment by nicksears** [comment]
@nophead You get good measurements with slic3r and one perimeter? A single outline definitely has the biggest impact IMO. Could we get the details of your setup? (slic3r ini settings) :smile:

**Comment by nophead** [comment]
No I get good measurements with Skeinforge with one perimeter.

**Comment by mrvn** [comment]
@DrLex0 polyholes only fixes the design flaws of using STL as intermediary format. You are designing in hole and polyhole will make it a polygon that sits outside the hole. That it also somewhat corrects for the printer dragging filament to the inside of corners is a lucky side effect of trying not to overwhelm the printer with too short segments.

**Comment by bubnikv** [comment]
@mrvn  We are going in circles. Instead of only commenting on the flaw of the current solution, please propose a better way to handle the holes.

**Comment by nophead** [comment]
So a better way:
   For each layer height, filament viscosity, nozzle size, print speed
measure the amount that filament lags behind the nozzle. That distance
makes the curve traced by the filament a tractrix. Then change the file
format from STL to something that represents curves as curves. Then have a
slicer that takes that format and moves the nozzle in an inverse tractrix
to get an accurate desired curve.

A lot of work and a lot of extra data required but polyholes are no work
and solve the problem of printing holes that things fit in regardless of
all the parameters I mentioned before. They don't need a new file format or
any slicer changes.

**Comment by DrLex0** [comment]
@bubnikv Yeah, I am starting to get déjà vus too.

I have tried polyholes and initially they also came out too small, so they are not a magic bullet solution for everything. It was only after I enforced narrower perimeter extrusion widths than Slic3r's auto setting, that they came out OK. However, rounder holes (with cos(1/n) correction) also come rather close to being accurate in that case. “External perimeters first” barely makes a difference, if there is an improvement in hole accuracy it is only marginal. I am starting to prefer it though, because it does a better job at avoiding small gaps between perimeters and infill.

I have also tested the too-many-segments theory by printing holes with high and low numbers of segments in both ABS and PLA, and found no evidence that it is valid in my case (printing from SD card on a 2016 FFCP). The printer does not slow down when printing the high-resolution holes, and there is no discernible difference with lower-resolution holes. This could be because I have set a resolution limit of at least 0.02 mm in all my print settings profiles, and going above 48 segments for a 3 mm hole pushes segment length below 0.2 mm.

**Comment by mrvn** [comment]
Yes, we are going in circles. Yes, polyholes are not a magic bullet. They only fix the problems on the design side and shortcomings of STL. It's the minimum you have to do. Obviously they are no different from round holes with cos(1/n) correction. Because that's exactly the correction polyholes does.

Now forget all about the polyholes and concentrate of the issues actually caused and relevant to slic3r. Namely modeling how the filament gets pushed and dragged inside the tool path. I'm with bubnikv there.

@DrLex0 Obviously you don't see the problem when printing from SD card. An SD card manages 1-10MiB/s, the serial porton the other hand only does 10-20KiB/s. That's 3 magnitudes slower. And a resolution limit of 0.02mm puts a limit there too.

**Comment by nophead** [comment]
>Because that's exactly the correction polyholes does.

No polyholes do two things. The cos(1/n) puts the sides of polygon on the
circle instead of the vertices. Since the sides are straight lines they are
accurately placed if only one outline is used. Reducing the number of
vertices to the minimum puts the corners as far outside the circle as
possible. That means that when the corners are rounded by the corner
cutting effect they are still outside the circle. That's how it works for
me anyway.

**Comment by ioquatix** [comment]
@mrvn could you solve the bandwidth problem by increasing buffer space on the device?

**Comment by bubnikv** [comment]
You would have to change the serial protocol to send and confirm multiple
lines at once. The bottleneck currently is the turn around time of the USB2
protocol.

On Aug 22, 2016 4:26 AM, "Samuel Williams" <notifications@github.com> wrote:

@mrvn <https://github.com/mrvn> could you solve the bandwidth problem by
increasing buffer space on the device?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/alexrj/Slic3r/issues/1588#issuecomment-241301957>, or mute
the thread

**Comment by mrvn** [comment]
@ioquatix Want to solder on some extra ram?

@bubnikv The serial protoctol already sends multiple lines and the firmware buffers it. Just a few lines though, like 3, since the problem is the lack of ram on the microcontrollers. It buffers just enough to calculate direction changes ahead of time so it can limit the acceleration and jerk.

**Comment by bubnikv** [comment]
> The serial protoctol already sends multiple lines and the firmware
buffers it. Just a few lines though, like 3, since the problem is the lack
of ram on the microcontrollers. It buffers just enough to calculate
direction changes ahead of time so it can limit the acceleration and jerk.

What firmware? What protocol?

AFAIK Marlin sends the g-code line by line, because there is no hardware
line control implemented.

**Comment by mrvn** [comment]
Any firmware should handle it. When sending gcode you can include a line number "Nxxxx" and the controler will respond with "OK xxxx" when it has handled that line. You can also include checksums and firmware is supposed to ask for resends of lines with checksum error or when line number are skipped (because they got lost due to lack of hardware flow control). 

Any firmware that has acceleration settings has to not only handle it but also has to read ahead. Without buffering and read ahead the tool would have to stop after every segment making a mokery of the whole thing. Marlin does this and things like Octoprint do send multiple lines (iirc 3) before waiting for OK replies.

**Comment by bubnikv** [comment]
> Marlin does this and things like Octoprint do send multiple lines (iirc
3) before waiting for OK replies.

That's exactly the problem. It waits for OK after each line (all right, it
may be 3). This turnaround takes 1ms.

**Comment by nophead** [comment]
Although the lines are sent one by one the G1 moves go into the planning buffer after the OK is sent, so as long as there are not too many segments per second the protocol does not slow it down. There is a limit though with USB and a much higher one with SD.

So printing too fast can cause the holes to shrink if the planning buffer runs dry. It also causes the filament to drag more, so corner cutting gets worse.

**Comment by nicksears** [comment]
OK, I've had a beef with this issue for a while now. I've worked on the flow math behind it before, but haven't come up with an elegant formula that could just say "increase internal radius by a factor of X".  The idea is that we're extruding a flattened torus, but it is pushing towards the inside more than the outside, so the volumetric midpoint is not the center of the nozzle.

The points on slic3r's manual are valid (http://manual.slic3r.org/troubleshooting/dimension-errors) but I don't think they account for the whole issue. Here's how I propose proving this is more than just a machine calibration issue and we need some kind of compensation algorithm:

1. Use a single perimeter to avoid multiple perimeters pushing inward (this happens for me even with outer perimeter first selected)
3. Use filament with very low variation, (Within 0.02-0.03mm) for very low flow irregularity which can cause layers to squish inward (and out)
3. Polygon approximations (polyholes) may help compensate for low resolution, but we should also just be able to export and slice at high resolution if this is truly the problem. We should use as high as practically feasible (0.005 is about the max for solidworks, and although it will make slicing slow, this may be reasonable for slic3r as well).
4. Finally, we can also avoid possible speed/stutter issues by printing fairly slow (30mm/s)

This should be easy to test without any physical changes to the printer. If you have a well-tuned/high quality printer, those should be the only possible issues, but for most printers there are also two other main considerations: Z-wobble and backlash.

5. Avoid Z-wobble by using a machine with precise, ballscrew z-axis, no perceptible z-wobble
6. Reduce backlash, completely tune out and test with dial guage within .001" (25um)

I'll try these if my first attempt is not successful with steps 1-4. I plan on trying a range of holes 3.0-3.5mm and see where an exact 3mm rod fits, but a 3mm bolt that usually measures ~2.9 would also pass in my book. I usually have to add 0.4-0.5mm for a through-hole fit, but, I've also used 0.25-0.35 for a tight fit. Any thoughts?

**Comment by ioquatix** [comment]
@nicksears It's a good idea. There must be some mathematical model for this that can be leveraged to do the right thing.

**Comment by nicksears** [comment]
This issue has been closed, despite being unresolved for some time. If anyone knows of a more active open issue, please let me know.

I have a few new thoughts on this. The majority of this issue when printing circles. I'm not saying it doesn't happen with internal squares, etc, but for the most part people have issues with curves.

The fact that polyholes can compensate for this error is a perfect indication of this. Polyholes proves that undersized holes result from curves, and you can mathematically determine how many external corners you need so that the internal rounding of those corners does not impinge upon the intended diameter. This would not work if internal flat measurements also had this issue.

This may be something many of you have thought of, but I've never thought of it this way. To me, this proves it's almost entirely a curve issue, not mechanical calibration.

**Comment by ioquatix** [comment]
I ended up using the following code across all my projects which makes the appropriate compensation:

https://github.com/ioquatix/zonestar-p802q-fan-duct/blob/8fde334e706c33159576af5fbea8cdbe1549466a/source/bolts.scad#L1-L13

It's worked well across a wide range of prints and printers (according to others who have printed the same models).
